src/services/shop/shop-router.ts

```diff
@@ -42,148 +43,423 @@ shopRouter.get(
+        role: Role.USER,
+        summary: "Returns qr code of users cart",
+        responses: {
+            [StatusCode.SuccessOK]: {
+                description: "QR code",
+                schema: OrderQRCodesSchema,
+            },
+            [StatusCode.ClientErrorNotFound]: {
+                description: "Shop Item DNE",
+                schema: ShopItemNotFoundErrorSchema,
+            },
+            [StatusCode.ClientErrorBadRequest]: {
+                description: "Not enough quantity in shop",
+                schema: ShopInsufficientFundsErrorSchema,
+            },
+            [StatusCode.ClientErrorPaymentRequired]: {
+                description: "User doesn't have enough points to purchase",
+                schema: ShopInsufficientFundsErrorSchema,
+            },
+            [StatusCode.ServerErrorInternal]: {
+                description: "Errors that should never happen",
+                schema: ShopInternalErrorSchema,
+            },
+        },
+    }),
+    async (req, res) => {
+        const { id: userId } = getAuthenticatedUser(req);
+
+        // Fetch user order
+        const userOrder = await Models.ShopOrder.findOne({ userId });
+        if (!userOrder) {
+            return res.status(StatusCode.ServerErrorInternal).send(ShopInternalError);
+        }
+
+        const { items, quantity } = userOrder;
+
+        // Fetch all shop items in one query
+        const shopItems = await Models.ShopItem.find({ itemId: { $in: items } });
+        const itemMap = new Map(shopItems.map((item) => [item.itemId, item]));
+
+        // Validate item availability
+        for (let i = 0; i < items.length; i++) {
+            const item = itemMap.get(items[i] ?? "");
+            if (!item) {
+                return res.status(StatusCode.ClientErrorNotFound).send(ShopItemNotFoundError);
+            }
+            const currentQuantity = quantity[i] ?? 0;
+            if (currentQuantity > item.quantity) {
+                return res.status(StatusCode.ClientErrorBadRequest).send(ShopInsufficientFundsError);
+            }
+        }
+
+        // Fetch user profile once
+        const profile = await Models.AttendeeProfile.findOne({ userId });
+        if (!profile) {
+            return res.status(StatusCode.ServerErrorInternal).send(ShopInternalError);
+        }
+
+        // Compute total cost
+        const totalPrice = items.reduce((sum, itemId, i) => {
+            const itemPrice = itemMap.get(itemId)?.price ?? 0; // Default to 0 if item price is not found
+            const itemQuantity = quantity[i] ?? 0; // Default to 0 if quantity is undefined
+            return sum + itemPrice * itemQuantity;
+        }, 0);
+
+        // Check if user has enough points
+        if (profile.points < totalPrice) {
+            return res.status(StatusCode.ClientErrorPaymentRequired).send(ShopInsufficientFundsError);
+        }
+
+        // Generate QR code
+        const qrCodeUrl = `hackillinois://shop?userId=${userId}`;
```

Comment:
This is not sufficient - use the user qr code generation. We need a qr that expires and can't just be generated by another user. With this logic I could make a qr for any arbitrary user and redeem their prize as them.